araproc/analysis/csw.py

```diff
@@ -0,0 +1,479 @@
+    useful_event, 
+    solution : int ,
+    polarization : int,
+    excluded_channels = None,
+    which_distance : str = "distant",
+    return_rolled_wfs : bool = False
+):
+    """
+    From a given `useful_event`, return the coherently summed waveform (CSW)
+    """
+
+    # Pull waveforms from the dataset
+    waveforms = data.get_waveforms(useful_event=useful_event, which_traces='filtered')
+
+    # Make a list of channels to analyze
+    channels_to_csw = _get_channels_to_csw(excluded_channels, data, polarization)
+
+    # Calculate the arrival delays in each channel relative to some 
+    #   reference channel
+    reference_ch = _get_channel_with_max_V(waveforms, channels_to_csw)
+    arrival_delays = _get_arrival_delays(
+        data, waveforms, channels_to_csw, excluded_channels, reference_ch,
+        which_distance, polarization, solution)
+    
+    # Shift each waveform by the arrival delay then interpolate each waveform 
+    #   so it exists within an array that stretches from the earliest wf 
+    #   time to the last wf time. Waveform values are set to `nan` at all 
+    #   times for which it does not exist. These will be trimmed later.
+    big_times = _get_catchall_time_array(waveforms, channels_to_csw, arrival_delays)
+    interpolated_wfs = {
+        ch_ID: np.full(len(big_times), -123456.0) for ch_ID in channels_to_csw}
+    for ch_ID in channels_to_csw:
+        interpolated_wfs[ch_ID] = _shift_wf(
+            *wu.tgraph_to_arrays(waveforms[ch_ID]), 
+            arrival_delays[ch_ID], big_times
+        )
+
+    # Roll the non-nan part of the waveforms around the preselected reference 
+    #   channel & build the CSW
+    roll_shifts = _get_shifts(interpolated_wfs, reference_ch)
+    rolled_wfs = {ch_ID: np.full(len(big_times), -123456.0) for ch_ID in channels_to_csw}
+    csw = np.zeros((1,len(big_times)))
+    for c, ch_ID in enumerate(channels_to_csw):
+        rolled_wfs[ch_ID] = _roll_wf(interpolated_wfs[ch_ID], roll_shifts[ch_ID], big_times)
+        # rolled_wfs[ch_ID] = interpolated_wfs[ch_ID] ## Uncomment to use unrolled wfs in csw
+        csw = np.nansum( np.dstack( (csw[0], rolled_wfs[ch_ID]) ), axis=2) 
+
+    # Un-nest the csw. csw.shape was (1,len(big_times)) but is now len(big_times)
+    csw = csw[0]
+
+    # Trim the CSW of leading and trailing nan values. Trim time array accordingly.
+    idx_non_pad_values = csw != 0.0
+    start_idx = np.where( 
+        idx_non_pad_values.any(), idx_non_pad_values.argmax(), -123456)
+    end_idx = np.where(
+        idx_non_pad_values.any(),
+        len(csw) - np.flip(idx_non_pad_values).argmax() - 1,
+        -123456
+    )
+    final_csw = csw[ start_idx : end_idx+1 ]
+    final_times = big_times[ start_idx : end_idx+1 ]
+
+    # Return requested data
+    if return_rolled_wfs: 
+        return final_times, final_csw, big_times, rolled_wfs
+    else:
+        return final_times, final_csw
+
+def get_csw_snr(csw : np.ndarray):
```

Comment:
I think we can just make this a wrapper around `snr.get_snr(csw)`. That way we are using the same snr definition throughout. :)

Reply:
Done!